### PR TITLE
[CBRD-24735] When attempting 'Multiple Key Ranges Optimization (MRO)', An internal error occurs because the '1=1' predicate (data filer) generated by constant folding is not removed

### PR DESCRIPTION
**Pull request by youngjinj**
http://jira.cubrid.org/browse/CBRD-24735

When the pt_semantic_type function runs after rewriting the query, it runs pt_where_type after the fold constant to see if there are any predicates that can be removed.

In the query below, the `1=1` predicate generated after rewriting the query should be removed.
```
pt_semantic_check: select ... from t0 inner join t1 t on t0.c1=t.c1 and t0.c1=1         where t.c1=1                                    order by 1 limit 1;
mq_translate:      select ... from t0 inner join t1 t                                   where t.c1=1 and t0.c1=t.c1 and t0.c1=1         order by 1 limit 1;
mq_optimize:       select ... from t0 inner join t1 t                                   where t.c1=1 and t0.c1=t.c1 and t0.c1=1 and 1=1 order by 1 limit 1;
    qo_reduce_equality_terms: t.c1=1 and t0.c1=t.c1 and t0.c1=1 (join_term_list: t0.c1=t.c1)
    qo_reduce_equality_terms: t.c1=1 and 1=1 and t0.c1=1        (join_term_list: t0.c1=t.c1)
    qo_reduce_equality_terms: t.c1=1 and t0.c1=t.c1 and t0.c1=1 and t0.c1=t.c1
pt_semantic_type:  select ... from t0 inner join t1 t on t0.c1=t.c1 and t0.c1=1 and 1=1 where t.c1=1                                    order by 1 limit 1;
```

### CODE REVIEW

File: src/parser/type_checking.c [hunk 1]
```diff
@@ -20358,6 +20358,8 @@ pt_semantic_type (PARSER_CONTEXT * parser, PT_NODE * tree, SEMANTIC_CHK_INFO * s
   tree = parser_walk_tree (parser, tree, pt_eval_type_pre, sc_info_ptr, pt_eval_type, sc_info_ptr);
   /* do constant folding */
   tree = parser_walk_tree (parser, tree, pt_fold_constants_pre, NULL, pt_fold_constants_post, sc_info_ptr);
+  /* do type checking again */
+  tree = parser_walk_tree (parser, tree, pt_eval_type_pre, sc_info_ptr, pt_eval_type, sc_info_ptr);
```
Comment (ctshim) [inline src/parser/type_checking.c]: It seems that error handling is needed.

 if (pt_has_error (parser))
    {
      return  NULL;
    }
    /* do type checking again */
  tree = parser_walk_tree (parser, tree, pt_eval_type_pre, sc_info_ptr, pt_eval_type, sc_info_ptr);

**Review by ctshim — APPROVED** [review]

**Review by hyunikn — APPROVED** [review]

**Review by beyondykk9 — APPROVED** [review]

**Comment by hgryoo** [comment]
Can I know where the node corresponding to '1=1' is generated?

**Comment by youngjinj** [comment]
Where does '1=1' come from?

```
db_compile_statement
  -> db_compile_statement_local
       -> mq_translate
            -> mq_translate_helper
                 -> mq_optimize
                      -> qo_optimize_queries
                           -> qo_move_on_clause_of_explicit_join_to_where_clause
                           -> qo_reduce_equality_terms_post
                                -> qo_reduce_equality_terms
                                     -> pt_lambda_with_arg
                                          -> pt_lambda_check_reduce_eq
                                          -> pt_lambda_node
                      -> qo_optimize_queries_post
```

When running the qo_reduce_equality_terms function, copy the join predicate into `join_term_list`.
In the reproduced scenario, `[dba.t0].c1=t.c1` is copied to `join_term_list`.

When the pt_lambda_with_arg function is executed, `wherep` has `t.c1=1 and [dba.t0].c1=t.c1 and [dba.t0].c1=1`.
When the pt_lambda_node function is executed, `[dba.t0].c1=t.c1` becomes `1=1` by `t.c1=1` and `[dba.t0].c1=1`.

After the execution of the pt_lambda_with_arg function is finished, `wherep` has `t.c1=1 and 1=1 and [dba.t0].c1=1`.
And add `join_term_list` here.

Then it becomes `t.c1=1 and 1=1 and [dba.t0].c1=1 and [dba.t0].c1=t.c1` as if `1=1` was added.

File: src/parser/type_checking.c (hunk 1 above)
Comment (youngjinj) [reply src/parser/type_checking.c]: Keep the existing error-checking code.
And I added code to check errors for the added code.

File: src/parser/type_checking.c [hunk 2]
```diff
@@ -20363,6 +20401,19 @@ pt_semantic_type (PARSER_CONTEXT * parser, PT_NODE * tree, SEMANTIC_CHK_INFO * s
       tree = NULL;
     }
 
+  /* When pt_lambda_node is executed in mq_optimize, a removable predicate like '1=1' is generated.
```
Comment (shparkcubrid) [inline src/parser/type_checking.c]: pt_lambda_node() is a general function to exchange pt_node. qo_reduce_equality_terms() is a more specific function.

**Review by shparkcubrid — APPROVED** [review]

**Review by hgryoo — APPROVED** [review]

**Review by beyondykk9 — APPROVED** [review]

File: src/parser/type_checking.c (hunk 2 above)
Comment (youngjinj) [reply src/parser/type_checking.c]: I will change the function name mentioned in the comments to qo_reduce_equality_terms.

**Pull request merged**